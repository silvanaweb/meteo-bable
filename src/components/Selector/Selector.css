.LocationSelector__list {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
}

.LocationSelector__item {
  margin: 0 0.4rem 0.8rem;
}
