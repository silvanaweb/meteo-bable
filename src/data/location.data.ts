// This file provide the location DataCue, that means city and position.
// If the cities were a user input, we would have used an api call to a coordinate locator.
// Because we know the cities, we can hardcode their coordinates in a dictionary.
// We also use union types as helpers

export type Places = "london" | "paris" | "newyork" | "singapore" | "sydney";
export type Location = {
  name: string;
  latitude: number;
  longitude: number;
};
export type LocationsMap = { [key in Places]: Location };

const locations: LocationsMap = {
  london: {
    name: "London",
    latitude: 51.507351,
    longitude: -0.127758
  },
  paris: {
    name: "Paris",
    latitude: 48.856613,
    longitude: 2.352222
  },
  newyork: {
    name: "New York",
    latitude: 40.712776,
    longitude: -74.005974
  },
  singapore: {
    name: "Singapore",
    latitude: 1.28944,
    longitude: 103.849983
  },
  sydney: {
    name: "Sydney",
    latitude: -33.86882,
    longitude: 151.20929
  }
};

export function getLocations() {
  return { ...locations };
}
