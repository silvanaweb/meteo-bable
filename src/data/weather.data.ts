// This file provide the weather info. It fetches them from an API call.

export type WeatherInfo = {
  location: string;
  summary: string;
  temperature: number;
  icon: string;
};

export function getWeatherFromLocation(
  latitude: number,
  longitude: number
): Promise<WeatherInfo> {
  // simulate API call
  return new Promise(resolve => {
    setTimeout(() => {
      resolve({
        location: "London",
        summary: "Rainy days",
        temperature: 20,
        icon: "rain"
      });
    }, 2000);
  });
}
