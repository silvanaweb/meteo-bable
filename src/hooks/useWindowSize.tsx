import { useEffect, useState } from "react";

export type WinSize = {
	width: number;
	height: number;
};

// create a reusable custom hook that fetch the window height and width on resize
const useWindowSize: () => WinSize = () => {
	const [size, setSize] = useState<WinSize>({ width: window.innerWidth, height: window.innerHeight });

	useEffect(() => {
		const handleResize = () => setSize({ width: window.innerWidth, height: window.innerHeight });
		window.addEventListener("resize", handleResize);
		// remove the listener on component unmount
		return () => {
			window.removeEventListener("resize", handleResize);
		};
	}, []);

	return size;
};

export { useWindowSize };
