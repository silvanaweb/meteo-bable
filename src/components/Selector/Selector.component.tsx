import React, { useEffect } from "react";
import Select from "react-select";
import { useWindowSize, WinSize } from "../../hooks/useWindowSize";
import { CheckButton } from "../CheckButton/CheckButton.component";
import { When } from "../Helpers/When";
import "./Selector.css";
export type SelectorOption = { value: string; label: string };

export type SelectorProps = {
  defaultOption?: SelectorOption;
  loading?: boolean;
  onChange: (option: any) => void;
  options: Array<SelectorOption>;
  title?: string;
};
const Selector: React.FC<SelectorProps> = ({
  defaultOption = null,
  loading = false,
  onChange,
  options,
  title
}) => {
  const [selectedOption, setSelectedOption] = React.useState<SelectorOption>();
  const winSize: WinSize = useWindowSize();

  useEffect(() => {
    if (defaultOption) {
      const aOption = options.find(x => x.value === defaultOption.value);
      if (aOption) {
        setSelectedOption(aOption);
        onChange({ ...aOption });
      }
    }
  }, [options]);

  const handleSelectedOption = (sOption: any) => {
    setSelectedOption(sOption);
    onChange({ ...sOption });
  };
  return (
    <div className="content-container">
      <When condition={!!title}>
        <h3 className="text-align--center">{title}</h3>
      </When>

      {// according to the size of the screen, it is using a Select or Check Buttons
      winSize.width < 600 ? (
        <div>
          <Select
            isDisabled={loading}
            onChange={handleSelectedOption}
            options={options}
            placeholder="Select a location"
            value={selectedOption}
          />
        </div>
      ) : (
        <div className="LocationSelector__list">
          {options.map((option: SelectorOption, idx: number) => (
            <div key={`genre.value${idx}`} className="LocationSelector__item">
              <CheckButton
                checked={
                  !!selectedOption && selectedOption.value === option.value
                }
                disabled={loading}
                name={option.value}
                onChange={() => handleSelectedOption(option)}
                title={option.label}
              />
            </div>
          ))}
        </div>
      )}
    </div>
  );
};

export { Selector };
