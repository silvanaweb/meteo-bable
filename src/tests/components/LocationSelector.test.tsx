import { shallow, ShallowWrapper } from "enzyme";
import toJSON from "enzyme-to-json";
import React from "react";
import { Selector } from "../../components/Selector/Selector.component";
import { optionsMock } from "../fixtures/location";
import { Location } from "../../data/location.data";
import * as WindowSizeHook from "../../hooks/useWindowSize";

let wrapper: ShallowWrapper;
let onChange: (location: Location) => void;

jest.mock("../../hooks/useWindowSize");

describe("Selector", () => {
  let useEffect: jest.SpyInstance;

  const mockUseEffect = () => {
    useEffect.mockImplementationOnce(f => f());
  };

  beforeEach(() => {
    useEffect = jest.spyOn(React, "useEffect");

    onChange = jest.fn();

    // mock the custom hook to have a desktop size
    const mockedWindowSizeHook = WindowSizeHook as jest.Mocked<
      typeof WindowSizeHook
    >;
    mockedWindowSizeHook.useWindowSize.mockImplementation(() => {
      return { width: 700, height: 700 };
    });

    mockUseEffect();
    mockUseEffect();
    mockUseEffect();

    wrapper = shallow(<Selector options={optionsMock} onChange={onChange} />);
  });

  test("should render Selector correctly", () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  test("should handle onChange when location is selected", () => {
    wrapper
      .find("CheckButton")
      .first()
      .dive()
      .find(`button`)
      .simulate("click", {
        stopPropagation: () => {}
      });

    expect(onChange).toHaveBeenLastCalledWith({
      ...optionsMock[0]
    });
  });
});
