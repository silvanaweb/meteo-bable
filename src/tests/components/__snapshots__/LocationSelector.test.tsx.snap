// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selector should render Selector correctly 1`] = `
<div
  className="content-container"
>
  <When
    condition={false}
  >
    <h3
      className="text-align--center"
    />
  </When>
  <div
    className="LocationSelector__list"
  >
    <div
      className="LocationSelector__item"
      key="genre.value0"
    >
      <CheckButton
        checked={false}
        disabled={false}
        name="london"
        onChange={[Function]}
        title="London"
      />
    </div>
    <div
      className="LocationSelector__item"
      key="genre.value1"
    >
      <CheckButton
        checked={false}
        disabled={false}
        name="paris"
        onChange={[Function]}
        title="Paris"
      />
    </div>
    <div
      className="LocationSelector__item"
      key="genre.value2"
    >
      <CheckButton
        checked={false}
        disabled={false}
        name="newyork"
        onChange={[Function]}
        title="New York"
      />
    </div>
    <div
      className="LocationSelector__item"
      key="genre.value3"
    >
      <CheckButton
        checked={false}
        disabled={false}
        name="singapore"
        onChange={[Function]}
        title="Singapore"
      />
    </div>
    <div
      className="LocationSelector__item"
      key="genre.value4"
    >
      <CheckButton
        checked={false}
        disabled={false}
        name="sydney"
        onChange={[Function]}
        title="Sydney"
      />
    </div>
  </div>
</div>
`;
